src/networking.rs

```diff
@@ -1,28 +1,128 @@
+            ));
+            let broadcast_outbound_addr = SocketAddr::V4(SocketAddrV4::new(
+                Ipv4Addr::new(255, 255, 255, 255),
+                *broadcast_port,
+            ));
+            let broadcast_raw_sock: std::net::UdpSocket = {
+                let broadcast_sock =
+                    Socket::new(Domain::IPV4, Type::DGRAM, Some(Protocol::UDP)).unwrap();
+                broadcast_sock.set_broadcast(true).unwrap();
+                broadcast_sock.set_nonblocking(true).unwrap();
+                broadcast_sock.set_reuse_address(true).unwrap();
+                broadcast_sock.set_reuse_port(true).unwrap();
+                broadcast_sock
+                    .bind(&SockAddr::from(broadcast_inbound_addr))
+                    .expect("Failed to bind for broadcast");
+                broadcast_sock.into()
+            };
+
+            // Unlike IPv6, we can use a single socket for both sending and receiving broadcasts,
+            // but consistency for consuming code, it makes sense to return two here as well.
+            let broadcast_in_sock = UdpSocket::from_std(broadcast_raw_sock.try_clone()?)?;
+            let broadcast_out_sock = UdpSocket::from_std(broadcast_raw_sock)?;
+
+            Ok((
+                broadcast_in_sock,
+                broadcast_out_sock,
+                broadcast_outbound_addr,
+            ))
+        }
+        IpAddr::V6(_) => {
+            // IPv6:
+            // - for listening to broadcasts, we have to join a multicast IP address.
+            // - for sending broadcast messages, we have to explicitly tell the socket which
+            //   network interface to use, otherwise it won't have a route to the multicast IP.
+
+            let Some(interface_idx) = get_interface_number(interface) else {
+                return Err(KaboodleError::UnableToFindInterfaceNumber);
+            };
+
+            // All IPv6 IP addresses in the ff00::/8 prefix are multicast addresses. The first 16
+            // bits of the IP indicate the "multicast group", and the remaining 112 bits are the
+            // "group ID". A prefix of 0xff02 indicates link-local multicast, where "link-local"
+            // means the traffic won't pass beyond the local router.
+            // Broadly speaking, we can use whatever group ID we like, so long as it doesn't
+            // conflict with any well-known services.
+            // See https://www.ciscopress.com/articles/article.asp?p=2803866&seqNum=5 for more
+            // information about IPv6 multicast addresses.
+            let broadcast_ip_addr = Ipv6Addr::new(0xff02, 0, 0, 0, 0, 0, 0x1213, 0x1989);
+
+            let broadcast_socket_addr =
+                SocketAddr::new(IpAddr::V6(broadcast_ip_addr), *broadcast_port);
+            let broadcast_in_sock = {
+                let sock = Socket::new(Domain::IPV6, Type::DGRAM, Some(Protocol::UDP))?;
+                sock.join_multicast_v6(&broadcast_ip_addr, interface_idx)?;
```

Comment:
This interface number stuff took me a whole day to figure out! You have to pass in the number of the interface to use for sending broadcast messages; without this, you end up with “no route to host” errors.

What's an interface number? I'm glad you asked. Every network interface on your system has u32 associated with it; `ifconfig -v | grep index` will let you see yours.

I ended up having to pull in the network-interface crate to get the interface number for a given interface.